### PR TITLE
* SCP-2667 Handle Connection reset by peer errors for MongoDB from PAPI

### PR DESCRIPTION
**Pull request by knapii-developments**
This PR adds handling for ```AutoReconnect: Connection reset by peer``. When the exception is encountered, the mongo command is retried up to 5 times prior to raising the exception.  A test has been added to both test_ingest and test_expressions_files.

This implementation allows at least 20 large studies to be uploaded without hitting the error.

### CODE REVIEW

File: ingest/mongo_connection.py [hunk 1]
```diff
@@ -29,3 +34,36 @@ def __init__(self):
...
+                return mongo_op_func(*args, **kwargs)
+            except AutoReconnect as e:
+                if attempt < MAX_AUTO_RECONNECT_ATTEMPTS - 1:
+                    wait_t = 0.5 * pow(2, attempt)  # exponential back off
+                    logging.warning(
+                        "PyMongo auto-reconnecting... %s. Waiting %.1f " "seconds.",
+                        str(e),
+                        wait_t,
+                    )
+                    time.sleep(wait_t)
```
Comment (eweitz) [inline ingest/mongo_connection.py:61]:
```suggestion
                    wait_time = 0.5 * pow(2, attempt)  # exponential back off
                    logging.warning(
                        "PyMongo auto-reconnecting... %s. Waiting %.1f " "seconds.",
                        str(e),
                        wait_time,
                    )
                    time.sleep(wait_time)
```

File: ingest/mongo_connection.py [hunk 2]
```diff
@@ -29,3 +34,36 @@ def __init__(self):
...
+    @functools.wraps(mongo_op_func)
+    def wrapper(*args, **kwargs):
+        for attempt in range(MAX_AUTO_RECONNECT_ATTEMPTS):
+            try:
+                return mongo_op_func(*args, **kwargs)
+            except AutoReconnect as e:
+                if attempt < MAX_AUTO_RECONNECT_ATTEMPTS - 1:
+                    wait_t = 0.5 * pow(2, attempt)  # exponential back off
+                    logging.warning(
+                        "PyMongo auto-reconnecting... %s. Waiting %.1f " "seconds.",
```
Comment (eweitz) [inline ingest/mongo_connection.py:57]:
```suggestion
                        "PyMongo auto-reconnecting... %s. Waiting %.1f seconds.",
```

File: ingest/expression_files/expression_files.py [hunk 3]
```diff
@@ -16,17 +16,16 @@
...
     from ingest_files import DataArray
     from monitor import setup_logger
-    from mongo_connection import MongoConnection
+    from mongo_connection import MongoConnection, graceful_auto_reconnect
 except ImportError:
     # Used when importing as external package, e.g. imports in single_cell_portal code
     from ..ingest_files import DataArray
-    from ..monitor import setup_logger
-    from ..mongo_connection import MongoConnection
+    from ..monitor import setup_logger, graceful_auto_reconnect
```
Comment (jlchang) [inline ingest/expression_files/expression_files.py:27]: I can't see why an import of graceful_auto_reconnect was added from monitor - is this needed?

**Comment by knapii-developments** [comment]
> Looks good. Neither of my suggested code changes are blocking.
> 
> A blocking request: please either re-run the scale test that prompted this work, or create a ticket for that for next sprint. The [stress test summary](https://docs.google.com/spreadsheets/d/1Q9RyacYLWC5Loe5MYgyJgT-sUj7E2PR2RlrErnS1Zf8/edit#gid=146459414) reports that 17 of 20 concurrent datasets ingested succesfully, and 3 failed.
> 
> Without re-running the same test (I think `python ingest_stress_test.py --token=$ACCESS_TOKEN -n 20 --data-set large` [per here](https://docs.google.com/document/d/11BZQKstFgENqGIHaFuOOrHpYOedXXS86JUt1bNngTdk/edit#), using an image incorporating these changes), I don't think we can know whether these changes fix the underlying issue.

I was able to upload 20 complete studies without hitting the error on my developers instance. This concludes that my implementation solves the issue.

File: tests/test_expression_files.py [hunk 4]
```diff
@@ -113,3 +115,12 @@ def test_insert(self):
         self.assertRaises(
             Exception, GeneExpression.insert, docs, "collection", client_mock
         )
+
+        # Test exponential back off
+        client_mock["collection"].insert_many.side_effect = AutoReconnect
```
Comment (bistline) [inline tests/test_expression_files.py:120]: Slick

File: ingest/expression_files/expression_files.py (hunk 3 above)
Comment (knapii-developments) [reply ingest/expression_files/expression_files.py:27]: 8ceccfc2

Comment (knapii-developments) [reply ingest/expression_files/expression_files.py]: Thanks for catching this!

**Pull request merged**